File: backend/app/agents/summary_agent.py
from __future__ import annotations

from app.agents.agentscope_runtime import run_structured_agent_sync
from app.agents.common import build_highlights, contains_chinese, extract_keywords, normalize_whitespace, summarize_text
from app.agents.context import AgentExecutionContext
from app.agents.structured_models import SummaryStructuredOutput


def _build_optimized_title(news_title: str, translated_title: str | None, keywords: list[str]) -> str:
    if translated_title and translated_title.strip():
        return normalize_whitespace(translated_title)
    if contains_chinese(news_title):
        return normalize_whitespace(news_title)
    if keywords:
        return f"{keywords[0]} | {news_title}"
    return normalize_whitespace(news_title)


def _build_summary_user_prompt(
    context: AgentExecutionContext,
    hotspot_result: dict[str, object],
    classification_result: dict[str, object],
) -> str:
    news = context.news
    return "\n".join(
        [
            "请根据下面这条 AI 新闻生成优化标题、中文摘要和要点，只输出结构化结果，不要额外解释。",
            f"标题：{news.title}",
            f"来源：{news.source}",
            f"分类：{classification_result.get('category') or news.category or '未分类'}",
            f"热度分值：{hotspot_result.get('score', news.hot_score)}",
            f"标签：{', '.join(classification_result.get('tags') or []) or '无'}",
            f"原始摘要：{news.summary or '无'}",
            "正文：",
            news.content or "",
        ]
    )


def run_summary_agent(
    context: AgentExecutionContext,
    hotspot_result: dict[str, object],
    classification_result: dict[str, object],
) -> dict[str, object]:
    runtime = context.runtime
    if runtime is not None:
        agent_result = run_structured_agent_sync(
            runtime=runtime,
            agent_name="SummaryAgent",
            system_prompt=context.prompt_bundle.get("agent_summary_prompt", ""),
            user_prompt=_build_summary_user_prompt(context, hotspot_result, classification_result),
            structured_model=SummaryStructuredOutput,
        )
        if agent_result is not None:
            return {
                **agent_result,
                "model_name": context.model_name,
                "prompt_version": context.prompt_version,
            }

    news = context.news
    text = news.content or news.summary or news.title
    keywords = list(classification_result.get("keywords") or extract_keywords(f"{news.title} {text}"))
    summary = summarize_text(text)
    highlights = build_highlights(text, keywords)
    optimized_title = _build_optimized_title(news.title, news.translated_title, keywords)
    if not summary:
        summary = normalize_whitespace(news.summary or news.title)
    if len(highlights) < 3:
        highlights.append(f"标题：{optimized_title}")

    return {
        "title": optimized_title,
        "summary": summary,
        "highlights": highlights[:4],
        "model_name": context.model_name,
        "prompt_version": context.prompt_version,
        "source_summary": normalize_whitespace(news.summary or "")[:180],
    }

File: backend-ts/src/agents/summary.agent.ts
/**
 * SummaryAgent - Summary generation agent
 * Generates optimized title, concise summary and key highlights
 *
 * Pattern: AI-first rule-fallback
 */

import { Injectable, Logger } from '@nestjs/common';
import { z } from 'zod';

import { AgentScopeRuntimeService } from './agents.runtime';
import type { AgentExecutionContext, ClassificationOutput, SummaryOutput } from './agents.types';
import {
  buildHighlights,
  extractKeywords,
  normalizeWhitespace,
  summarizeText,
} from './agents.common';

// ===== Zod Schema =====

export const SummarySchema = z.object({
  title: z.string(),
  summary: z.string(),
  highlights: z.array(z.string()).default([]),
  source_summary: z.string().default(''),
});

@Injectable()
export class SummaryAgent {
  private readonly logger = new Logger(SummaryAgent.name);

  constructor(private readonly runtime: AgentScopeRuntimeService) {}

  private buildUserPrompt(
    ctx: AgentExecutionContext,
    classificationResult: ClassificationOutput,
  ): string {
    const news = ctx.news;
    return [
      'Generate an optimized title, one-line summary and key highlights for this AI news. Output structured JSON only.',
      `Title: ${news.title}`,
      `Source: ${news.source}`,
      `Category: ${classificationResult.category}`,
      `Tags: ${classificationResult.tags.join(', ')}`,
      `Keywords: ${classificationResult.keywords.join(', ')}`,
      `Language: ${news.language}`,
      `Original Summary: ${news.summary || 'N/A'}`,
      'Content:',
      news.content || '',
    ].join('\n');
  }

  async run(
    ctx: AgentExecutionContext,
    classificationResult: ClassificationOutput,
  ): Promise<SummaryOutput> {
    const aiResult = await this.runtime.callStructured<SummaryOutput>({
      agentName: 'SummaryAgent',
      systemPrompt: ctx.prompt_bundle.agent_summary_prompt || '',
      userPrompt: this.buildUserPrompt(ctx, classificationResult),
      schema: SummarySchema,
    });

    if (aiResult) {
      this.logger.debug(`SummaryAgent AI result: title=${aiResult.title.slice(0, 50)}`);
      return aiResult;
    }

    return this.fallback(ctx, classificationResult);
  }

  private fallback(
    ctx: AgentExecutionContext,
    classificationResult: ClassificationOutput,
  ): SummaryOutput {
    const news = ctx.news;
    const content = news.content || news.summary || '';
    const keywords = extractKeywords(`${news.title} ${content}`);

    return {
      title: news.title.length > 120 ? news.title.slice(0, 117) + '...' : news.title,
      summary: summarizeText(content),
      highlights: buildHighlights(content || '', keywords),
      source_summary: normalizeWhitespace(news.summary) || '',
    };
  }
}
